File: java/src/main/java/dev/envoice/InvoiceResult.java
package dev.envoice;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Base64;
import java.util.List;

/**
 * Result of an invoice generation
 */
public sealed interface InvoiceResult {

    boolean isSuccess();

    /**
     * Successful invoice generation
     */
    record Success(
        String pdfBase64,
        String filename,
        Types.ValidationResult validation,
        Types.AccountInfo account
    ) implements InvoiceResult {

        @Override
        public boolean isSuccess() {
            return true;
        }

        /**
         * Save the PDF to a file
         */
        public void savePdf(String path) throws IOException {
            savePdf(Path.of(path));
        }

        /**
         * Save the PDF to a file
         */
        public void savePdf(Path path) throws IOException {
            Files.createDirectories(path.getParent());
            Files.write(path, toBytes());
        }

        /**
         * Get the PDF as bytes
         */
        public byte[] toBytes() {
            return Base64.getDecoder().decode(pdfBase64);
        }

        /**
         * Get the PDF as a data URL
         */
        public String toDataUrl() {
            return "data:application/pdf;base64," + pdfBase64;
        }
    }

    /**
     * Failed invoice generation
     */
    record Failure(List<Types.ValidationError> errors) implements InvoiceResult {

        @Override
        public boolean isSuccess() {
            return false;
        }
    }
}

File: csharp/src/InvoiceResult.cs
namespace Envoice;

/// <summary>Result of an invoice generation</summary>
public abstract record InvoiceResult
{
    public abstract bool IsSuccess { get; }
    public bool IsFailure => !IsSuccess;
}

/// <summary>Successful invoice generation</summary>
public record InvoiceSuccess : InvoiceResult
{
    public string PdfBase64 { get; }
    public string Filename { get; }
    public ValidationResult Validation { get; }
    public AccountInfo? Account { get; }

    public override bool IsSuccess => true;

    public InvoiceSuccess(string pdfBase64, string filename, ValidationResult validation, AccountInfo? account = null)
    {
        PdfBase64 = pdfBase64;
        Filename = filename;
        Validation = validation;
        Account = account;
    }

    /// <summary>Save the PDF to a file</summary>
    public async Task SavePdfAsync(string path, CancellationToken cancellationToken = default)
    {
        var directory = Path.GetDirectoryName(path);
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }
        await File.WriteAllBytesAsync(path, ToBytes(), cancellationToken);
    }

    /// <summary>Get the PDF as bytes</summary>
    public byte[] ToBytes() => Convert.FromBase64String(PdfBase64);

    /// <summary>Get the PDF as a data URL</summary>
    public string ToDataUrl() => $"data:application/pdf;base64,{PdfBase64}";
}

/// <summary>Failed invoice generation</summary>
public record InvoiceFailure : InvoiceResult
{
    public IReadOnlyList<ValidationError> Errors { get; }

    public override bool IsSuccess => false;

    public InvoiceFailure(IReadOnlyList<ValidationError> errors)
    {
        Errors = errors;
    }
}
